File: src/utils/validation.ts
/**
 * Internal validation helper functions.
 *
 * @internal
 * @packageDocumentation
 */

import * as crypto from "node:crypto";

/**
 * Parsed Telegram Mini App `initData` payload.
 */
export interface TelegramWebAppData {
  /**
   * Unique identifier for the Web App session.
   */
  query_id?: string;
  /**
   * Information about the current user.
   */
  user?: {
    id: number;
    is_bot?: boolean;
    first_name: string;
    last_name?: string;
    username?: string;
    language_code?: string;
    is_premium?: boolean;
    allows_write_to_pm?: boolean;
    photo_url?: string;
    [key: string]: unknown;
  };
  /**
   * Information about the user or bot who opened the Web App in direct chat.
   */
  receiver?: {
    id: number;
    is_bot?: boolean;
    first_name: string;
    last_name?: string;
    username?: string;
    [key: string]: unknown;
  };
  /**
   * Information about the chat from which the Web App was opened.
   */
  chat?: {
    id: number;
    type: string;
    title: string;
    username?: string;
    photo_url?: string;
    [key: string]: unknown;
  };
  /**
   * Type of the chat from which the Web App was opened.
   */
  chat_type?: string;
  /**
   * Global identifier for the chat instance.
   */
  chat_instance?: string;
  /**
   * Start parameter passed in the deep link.
   */
  start_param?: string;
  /**
   * Time in seconds after which a message can be sent via `answerWebAppQuery`.
   */
  can_send_after?: number;
  /**
   * Unix timestamp of when the Web App data was generated.
   */
  auth_date: number;
  /**
   * Signature hash for data integrity verification.
   */
  hash: string;
  /**
   * Raw key-value string dictionary from URLSearchParams.
   */
  raw: Record<string, string>;
}

/**
 * Options for validating Telegram Mini App `initData`.
 */
export interface ValidateWebAppDataOptions {
  /**
   * Maximum allowed age of the authentication data in seconds.
   * If provided and `auth_date` is older than `now - maxAgeSeconds`, validation returns false.
   * @defaultValue 86400 (24 hours)
   */
  maxAgeSeconds?: number;
}

/**
 * Asserts that a value is a non-empty string.
 *
 * @param value - The value to check.
 * @param name - The name of the parameter/field being validated (for error reporting).
 * @throws TypeError When the value is not a string or contains only whitespace.
 *
 * @example
 * ```ts
 * assertNonEmptyString(token, "token");
 * ```
 */
export function assertNonEmptyString(value: unknown, name: string): asserts value is string {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new TypeError(`${name} must be a non-empty string`);
  }
}

/**
 * Asserts that a value is a valid finite number.
 *
 * @param value - The value to check.
 * @param name - The name of the parameter/field being validated (for error reporting).
 * @throws TypeError When the value is not a number or is NaN.
 *
 * @example
 * ```ts
 * assertNumber(chatId, "chatId");
 * ```
 */
export function assertNumber(value: unknown, name: string): asserts value is number {
  if (typeof value !== "number" || Number.isNaN(value)) {
    throw new TypeError(`${name} must be a valid number`);
  }
}

/**
 * Validates bot token format.
 *
 * @param token - Token string to validate.
 * @throws TypeError When the token is empty, or non-empty but does not match the `<bot_id>:<secret>`
 * format (e.g. `"123456:ABC-DEF..."`). The literal `"TEST_TOKEN"` is always accepted for
 * use in tests and examples.
 */
export function validateToken(token: string): void {
  assertNonEmptyString(token, "token");
  if (token === "TEST_TOKEN") return;
  if (!/^\d+:[A-Za-z0-9_-]+$/.test(token)) {
    throw new TypeError(
      `token must match the Telegram bot token format "<bot_id>:<secret>" (e.g. "123456:ABC-DEF...")`,
    );
  }
}

/**
 * Validates Telegram Mini App `initData` signature using HMAC-SHA256 according to official Telegram specifications.
 *
 * @param initData - Raw URL-encoded query string received from `Telegram.WebApp.initData`.
 * @param botToken - The bot token used to verify data authenticity.
 * @param options - Optional validation settings such as `maxAgeSeconds`.
 * @returns `true` if the signature and expiration timestamp are valid; `false` otherwise.
 *
 * @example
 * ```ts
 * const isValid = validateWebAppData(req.headers["x-telegram-init-data"], botToken);
 * if (!isValid) {
 *   return res.status(401).send("Unauthorized");
 * }
 * ```
 */
export function validateWebAppData(
  initData: string,
  botToken: string,
  options: ValidateWebAppDataOptions = {},
): boolean {
  if (!initData || !botToken) return false;

  const params = new URLSearchParams(initData);
  const hash = params.get("hash");
  if (!hash) return false;

  const authDateStr = params.get("auth_date");
  if (!authDateStr) return false;

  const authDate = parseInt(authDateStr, 10);
  if (Number.isNaN(authDate)) return false;

  const maxAgeSeconds = options.maxAgeSeconds ?? 86400;
  if (maxAgeSeconds > 0) {
    const now = Math.floor(Date.now() / 1000);
    // Disallow timestamps from too far in the future (skew > 5 mins) or too old
    if (authDate > now + 300 || now - authDate > maxAgeSeconds) {
      return false;
    }
  }

  // Sort parameter pairs alphabetically (excluding "hash")
  const entries: string[] = [];
  params.forEach((val, key) => {
    if (key !== "hash") {
      entries.push(`${key}=${val}`);
    }
  });
  entries.sort();
  const dataCheckString = entries.join("\n");

  // secret_key = HMAC_SHA256("WebAppData", bot_token)
  const secretKey = crypto.createHmac("sha256", "WebAppData").update(botToken).digest();
  const calculatedHash = crypto.createHmac("sha256", secretKey).update(dataCheckString).digest("hex");

  if (calculatedHash.length !== hash.length) return false;

  return crypto.timingSafeEqual(Buffer.from(calculatedHash, "hex"), Buffer.from(hash, "hex"));
}

/**
 * Parses raw Telegram Mini App `initData` string into a structured {@link TelegramWebAppData} object.
 *
 * @param initData - Raw query string from `Telegram.WebApp.initData`.
 * @returns Parsed structured data object.
 *
 * @example
 * ```ts
 * const data = parseWebAppData(initDataString);
 * console.log(`User ID: ${data.user?.id}, Name: ${data.user?.first_name}`);
 * ```
 */
export function parseWebAppData(initData: string): TelegramWebAppData {
  const params = new URLSearchParams(initData);
  const raw: Record<string, string> = {};

  params.forEach((val, key) => {
    raw[key] = val;
  });

  const parseJsonField = <T>(key: string): T | undefined => {
    const val = params.get(key);
    if (!val) return undefined;
    try {
      return JSON.parse(val) as T;
    } catch {
      return undefined;
    }
  };

  const authDateStr = params.get("auth_date") || "0";
  const authDate = parseInt(authDateStr, 10) || 0;
  const canSendAfterStr = params.get("can_send_after");
  const canSendAfter = canSendAfterStr ? parseInt(canSendAfterStr, 10) : undefined;

  return {
    query_id: params.get("query_id") || undefined,
    user: parseJsonField(params.get("user") ? "user" : ""),
    receiver: parseJsonField(params.get("receiver") ? "receiver" : ""),
    chat: parseJsonField(params.get("chat") ? "chat" : ""),
    chat_type: params.get("chat_type") || undefined,
    chat_instance: params.get("chat_instance") || undefined,
    start_param: params.get("start_param") || undefined,
    can_send_after: Number.isNaN(canSendAfter) ? undefined : canSendAfter,
    auth_date: authDate,
    hash: params.get("hash") || "",
    raw,
  };
}

File: packages/go/pkg/utils/validation.go
package utils

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"net/url"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

// ErrInvalidToken is returned by ValidateToken when a bot token is empty or
// does not match the Telegram bot token format.
var ErrInvalidToken = errors.New("invalid bot token")

// tokenPattern matches the Telegram bot token format "<bot_id>:<secret>".
var tokenPattern = regexp.MustCompile(`^\d+:[A-Za-z0-9_-]+$`)

// defaultMaxAgeSeconds is the default maximum age accepted for Mini App
// initData authentication timestamps (24 hours).
const defaultMaxAgeSeconds = 86400

// maxClockSkewSeconds is how far in the future an auth_date may lie before
// it is rejected (5 minutes), to tolerate client clock skew.
const maxClockSkewSeconds = 300

// ValidateToken checks that token looks like a Telegram bot token of the
// form "<bot_id>:<secret>" (e.g. "123456:ABC-DEF..."). The literal
// "TEST_TOKEN" is always accepted for use in tests and examples. It returns
// an error wrapping ErrInvalidToken when the token is empty, blank, or
// malformed.
func ValidateToken(token string) error {
	if strings.TrimSpace(token) == "" {
		return fmt.Errorf("%w: token must be a non-empty string", ErrInvalidToken)
	}
	if token == "TEST_TOKEN" {
		return nil
	}
	if !tokenPattern.MatchString(token) {
		return fmt.Errorf(
			`%w: token must match the Telegram bot token format "<bot_id>:<secret>" (e.g. "123456:ABC-DEF...")`,
			ErrInvalidToken,
		)
	}
	return nil
}

// WebAppUser is a Telegram user as embedded in Mini App initData.
type WebAppUser struct {
	ID              int64  `json:"id"`
	IsBot           bool   `json:"is_bot,omitempty"`
	FirstName       string `json:"first_name"`
	LastName        string `json:"last_name,omitempty"`
	Username        string `json:"username,omitempty"`
	LanguageCode    string `json:"language_code,omitempty"`
	IsPremium       bool   `json:"is_premium,omitempty"`
	AllowsWriteToPM bool   `json:"allows_write_to_pm,omitempty"`
	PhotoURL        string `json:"photo_url,omitempty"`
}

// WebAppChat is a Telegram chat as embedded in Mini App initData.
type WebAppChat struct {
	ID       int64  `json:"id"`
	Type     string `json:"type"`
	Title    string `json:"title"`
	Username string `json:"username,omitempty"`
	PhotoURL string `json:"photo_url,omitempty"`
}

// WebAppData is the parsed representation of a Telegram Mini App initData
// query string.
//
// This is a framework helper for validated initData, not the Bot API object of
// the same name (Message.WebAppData in package types carries the payload a Web
// App button sent to the bot).
type WebAppData struct {
	// QueryID is the unique identifier of the Web App session.
	QueryID string
	// User is the current user, or nil when absent or unparseable.
	User *WebAppUser
	// Receiver is the user or bot the Web App was opened with in a direct
	// chat, or nil when absent or unparseable.
	Receiver *WebAppUser
	// Chat is the chat from which the Web App was opened, or nil when
	// absent or unparseable.
	Chat *WebAppChat
	// ChatType is the type of the chat from which the Web App was opened.
	ChatType string
	// ChatInstance is the global identifier of the chat instance.
	ChatInstance string
	// StartParam is the start parameter passed in the deep link.
	StartParam string
	// CanSendAfter is the number of seconds after which a message can be
	// sent via answerWebAppQuery; zero when absent.
	CanSendAfter int64
	// AuthDate is the Unix timestamp when the initData was generated.
	AuthDate int64
	// Hash is the signature used for integrity verification.
	Hash string
	// Raw holds every decoded key-value pair of the query string.
	Raw map[string]string
}

// ParseWebAppData parses a raw Telegram Mini App initData query string into
// a structured WebAppData object. JSON-encoded fields (user, receiver, chat)
// that fail to decode are left nil instead of failing the whole parse. It
// returns an error only when initData is not a valid URL-encoded query
// string.
func ParseWebAppData(initData string) (*WebAppData, error) {
	values, err := url.ParseQuery(initData)
	if err != nil {
		return nil, fmt.Errorf("invalid initData query string: %w", err)
	}

	data := &WebAppData{
		QueryID:      values.Get("query_id"),
		ChatType:     values.Get("chat_type"),
		ChatInstance: values.Get("chat_instance"),
		StartParam:   values.Get("start_param"),
		Hash:         values.Get("hash"),
		Raw:          make(map[string]string, len(values)),
	}
	for key, vals := range values {
		if len(vals) > 0 {
			data.Raw[key] = vals[len(vals)-1]
		}
	}

	if raw := values.Get("user"); raw != "" {
		var user WebAppUser
		if json.Unmarshal([]byte(raw), &user) == nil {
			data.User = &user
		}
	}
	if raw := values.Get("receiver"); raw != "" {
		var receiver WebAppUser
		if json.Unmarshal([]byte(raw), &receiver) == nil {
			data.Receiver = &receiver
		}
	}
	if raw := values.Get("chat"); raw != "" {
		var chat WebAppChat
		if json.Unmarshal([]byte(raw), &chat) == nil {
			data.Chat = &chat
		}
	}

	if raw := values.Get("auth_date"); raw != "" {
		if v, err := strconv.ParseInt(raw, 10, 64); err == nil {
			data.AuthDate = v
		}
	}
	if raw := values.Get("can_send_after"); raw != "" {
		if v, err := strconv.ParseInt(raw, 10, 64); err == nil {
			data.CanSendAfter = v
		}
	}

	return data, nil
}

// VerifyOption configures VerifyWebAppData.
type VerifyOption func(*verifyConfig)

type verifyConfig struct {
	maxAgeSeconds int64
}

// WithMaxAge sets the maximum allowed age of the initData auth_date in
// seconds. A value of zero or less disables the age check entirely.
// Defaults to 86400 (24 hours).
func WithMaxAge(seconds int64) VerifyOption {
	return func(c *verifyConfig) {
		c.maxAgeSeconds = seconds
	}
}

// VerifyWebAppData validates a Telegram Mini App initData signature using
// HMAC-SHA256 according to the official Telegram specification: the secret
// key is HMAC_SHA256("WebAppData", bot_token), and the expected hash is the
// hex-encoded HMAC_SHA256 of the alphabetically sorted, newline-joined
// key=value pairs (excluding the hash itself). It also rejects auth_date
// values older than maxAge (default 24 hours, see WithMaxAge) or more than
// 5 minutes in the future. It reports whether the data is authentic and
// fresh; any parse failure reports false.
func VerifyWebAppData(initData, botToken string, opts ...VerifyOption) bool {
	if initData == "" || botToken == "" {
		return false
	}

	cfg := verifyConfig{maxAgeSeconds: defaultMaxAgeSeconds}
	for _, opt := range opts {
		opt(&cfg)
	}

	values, err := url.ParseQuery(initData)
	if err != nil {
		return false
	}

	hash := values.Get("hash")
	if hash == "" {
		return false
	}

	authDateStr := values.Get("auth_date")
	if authDateStr == "" {
		return false
	}
	authDate, err := strconv.ParseInt(authDateStr, 10, 64)
	if err != nil {
		return false
	}

	if cfg.maxAgeSeconds > 0 {
		now := time.Now().Unix()
		// Disallow timestamps from too far in the future or too old.
		if authDate > now+maxClockSkewSeconds || now-authDate > cfg.maxAgeSeconds {
			return false
		}
	}

	// Sort parameter pairs alphabetically (excluding "hash").
	pairs := make([]string, 0, len(values))
	for key, vals := range values {
		if key == "hash" {
			continue
		}
		for _, val := range vals {
			pairs = append(pairs, key+"="+val)
		}
	}
	sort.Strings(pairs)
	dataCheckString := strings.Join(pairs, "\n")

	// secret_key = HMAC_SHA256("WebAppData", bot_token)
	secretKey := hmacSHA256([]byte("WebAppData"), []byte(botToken))
	calculatedHash := hex.EncodeToString(hmacSHA256(secretKey, []byte(dataCheckString)))

	if len(calculatedHash) != len(hash) {
		return false
	}
	got, err := hex.DecodeString(hash)
	if err != nil {
		return false
	}
	want, err := hex.DecodeString(calculatedHash)
	if err != nil {
		return false
	}
	return hmac.Equal(got, want)
}

// hmacSHA256 computes the HMAC-SHA256 of data under the given key.
func hmacSHA256(key, data []byte) []byte {
	mac := hmac.New(sha256.New, key)
	mac.Write(data)
	return mac.Sum(nil)
}
